sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java

```diff
@@ -0,0 +1,182 @@
+
+    private static final String STEP_AMOUNT = "datetime.step.amount";
+
+    private DateTimeFormatter datetimeFormatter;
+
+    private ChronoUnit stepUnit;
+
+    private int stepAmount;
+
+    private volatile boolean init;
+
+    @Getter
+    @Setter
+    private Properties properties = new Properties();
+
+    @Override
+    public String doSharding(final Collection<String> availableTargetNames, final PreciseShardingValue<Comparable<?>> shardingValue) {
+        checkInit();
+        return availableTargetNames.stream()
+                .filter(tableName -> tableName.endsWith(formatForDateTime(parseDateTimeForValue(shardingValue.getValue().toString()))))
+                .findFirst().orElseThrow(() -> new UnsupportedOperationException(
+                        String.format("failed to shard value %s, and availableTables %s", shardingValue, availableTargetNames)));
+    }
+
+    @Override
+    public Collection<String> doSharding(final Collection<String> availableTargetNames, final RangeShardingValue<Comparable<?>> shardingValue) {
+        checkInit();
+        boolean hasStart = shardingValue.getValueRange().hasLowerBound();
+        boolean hasEnd = shardingValue.getValueRange().hasUpperBound();
+        Set<String> tables = new HashSet<>();
+        if (!hasStart && !hasEnd) {
+            return availableTargetNames;
+        }
+        LocalDateTime start = hasStart
+                ? parseDateTimeForValue(shardingValue.getValueRange().lowerEndpoint().toString())
+                : parseDateTimeForValue(properties.getProperty(DEFAULT_LOWER));
+        LocalDateTime end = hasEnd
+                ? parseDateTimeForValue(shardingValue.getValueRange().upperEndpoint().toString())
+                : properties.getProperty(DEFAULT_UPPER) == null
+                ? LocalDateTime.now()
+                : parseDateTimeForValue(properties.getProperty(DEFAULT_UPPER));
+        LocalDateTime tmp = start;
+        while (!tmp.isAfter(end)) {
+            mergeTableIfMatch(tmp, tables, availableTargetNames);
+            tmp = tmp.plus(stepAmount, stepUnit);
+        }
+        mergeTableIfMatch(end, tables, availableTargetNames);
+        return tables;
+    }
+
+    private LocalDateTime parseDateTimeForValue(final String value) {
+        return LocalDateTime.parse(value.substring(0, properties.getProperty(DATE_TIME_FORMAT).length()), datetimeFormatter);
+    }
+
+    private String formatForDateTime(final LocalDateTime localDateTime) {
+        return localDateTime.format(DateTimeFormatter.ofPattern(properties.get(TABLE_SUFFIX_FORMAT).toString()));
+    }
+
+    private void mergeTableIfMatch(final LocalDateTime dateTime, final Collection<String> tables, final Collection<String> availableTargetNames) {
+        String suffix = formatForDateTime(dateTime);
+        availableTargetNames.parallelStream().filter(tableName -> tableName.endsWith(suffix)).findAny().map(tables::add);
+    }
+
+    private void checkInit() {
```

Comment:
why add sync block here？

Reply:
sync block here is used to ensure `verifyProperties()` method exactly execute once when multi-thread call `doSharding()`, inspired by Double-Checked Locking with Singleton